### PR TITLE
#[track_caller] in traits

### PR DESCRIPTION
**Pull request by anp**
Per https://github.com/rust-lang/rust/issues/47809#issuecomment-572791760, this allows the `#[track_caller]` attribute on trait methods.

Includes tests for `#[track_caller]` with:

* "regular" trait impls
* default trait impls
* "blanket-tracked" trait impls, where the annotation is in the trait definition and is inherited by "regular" impls of the trait

### CODE REVIEW

**Comment by anp** [comment]
cc @eddyb you mentioned wanting to have some particular tests around ReifyShim and the trait impl. I've added traits to the fn-ptr tests, take a look?

**Review by eddyb — APPROVED** [review]

**Comment by eddyb** [comment]
LGTM, as long as we want to do this.

**Comment by anp** [comment]
Cool! IIUC the [lang team approved this](https://github.com/rust-lang/rust/issues/47809#issuecomment-572791760).

File: src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs
Comment (eddyb) [inline src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs]: Btw we might want this to actually work and change the ABI to pass the caller!

File: src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs
Comment (eddyb) [inline src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs]: We might also want to error here if we want to reserve the right to add meaning to it?

**Comment by Centril** [comment]
Let's just confirm that perf doesn't regress; @bors try @rust-timer queue

File: src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs
Comment (Centril) [reply src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs]: Would that still apply if the ABI is e.g. `"C"`?

File: src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs
Comment (Centril) [reply src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs]: It does seem prudent to error for now, but I'm curious, do you have thoughts re. meaning?

File: src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs
Comment (eddyb) [reply src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs]: I think we'd want that to be an error (it should already be in general, e.g. if you're *defining* an `extern "C" fn`, i.e. exporting it).

File: src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs
Comment (eddyb) [reply src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs]: It could be inherited by `impl`s by default, for example.

Comment (Centril) [reply src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs]: That seems plausible, but on the other hand, what is the remedy users would have to disable that effect on specific impls then?

File: src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs
Comment (Centril) [reply src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs]: We do have an error without the Rust ABI when exporting:
```rust
#![feature(track_caller)]

#[track_caller]
extern "C" fn foo() {} //~ ERROR `#[track_caller]` requires Rust ABI
```

@anp I don't recall seeing anything about this in the reference PR (https://github.com/rust-lang/reference/pull/742, let's make sure we do note that restriction). Also, do we have a test checking this? (E.g. for "C" and some other random obscure ABI.)

In importing contexts we get `error[E0739]: attribute should be applied to function` right now, which is confusing, but at least there's an error. :)

File: src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs
Comment (eddyb) [reply src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs]: They can introduce a non-`#[track_caller]` wrapper, the same way e.g. `libcore` introduces `#[cold]` `#[inline(never)]` functions around panics (although those serve a different purpose).

File: src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs
Comment (eddyb) [reply src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs]: > In importing contexts we get `error[E0739]: attribute should be applied to function` right now, which is confusing, but at least there's an error. :)

This PR changes that to a warning (which is what we're commenting on).

Comment (anp) [reply src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs]: Yeah, let's promote to an error, cover it in the reference, and punt on ABI semantics questions?

File: src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs
Comment (anp) [reply src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs]: I'd be very interested in a version of track_caller that's automatically applied to all implementors of a method -- happy to promote warning to error if it creates room for that down the road.

**Comment by anp** [comment]
I'm not sure what typical variances are on perf.rlo these days, but those numbers don't look great :/.

File: src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs
Comment (Centril) [reply src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs]: Sounds good. We do have a test `error-with-invalid-abi.rs` :)

File: src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs
Comment (Centril) [reply src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs]: @anp If you want to follow up with such a PR we could also deal with it before stabilization if you like so that everything is more "complete" when we do stabilize.

**Comment by Centril** [comment]
I think that's a clear, but not too big regression.

**Comment by eddyb** [comment]
You can ignore `syn-opt`, there's an issue open about its nondeterminism.

<hr/>

The weird thing here is `-check` builds being affected by what should be a codegen change only, and especially "clean incremental" runs!

<hr/>

For "baseline incremental" runs the regression is usually `0.3%` or less except for:
* `helloworld-debug` at `0.6%`
* `cranelift-codegen-opt` at `0.7%`
* `token-stream-stress-debug` at `0.4%`

<hr/>

Going back to "clean incremental" runs, the only one over `1.0%` with a change in the number of queries is [`ripgrep-opt`](https://perf.rust-lang.org/detailed-query.html?commit=cb29a8883f4d99b4554b6c987c8e58194efe3d8e&base_commit=b0d5813fd7fe4696b6eb61acf473a7f7a5f9b5a0&benchmark=ripgrep-opt&run_name=clean%20incremental), all the other ones execute the same number of queries, just slower.

<hr/>

Also, many runs that differ in total time, don't differ as much (or at all?) in total query time, that's confusing.

The commonly regressed queries appear to be `incr_comp_load_dep_graph` and... `param_env`?

<hr/>

Hang on, could this entire thing be because of the effect of adding `#[track_caller]` to all of these `impl`s on the compiler code itself?

cc @Mark-Simulacrum any way we could run perf with a `stage1` build, or maybe only enable the `#[cfg_attr(not(bootstrap), track_caller)]` added in this PR for `stage2` `libstd`?

@anp In the meanwhile, maybe we can land the PR without the changed `impl`s? And then maybe figure out separately which ones are responsible for slowing down rustc?

**Comment by Mark-Simulacrum** [comment]
> any way we could run perf with a stage1 build, or maybe only enable the #[cfg_attr(not(bootstrap), track_caller)] added in this PR for stage2 libstd?

Stage 1 build will be quite hard I think. The latter, presuming @anp or whoever can rename the attribute to `#[cfg_attr(tracking_caller, track_caller)]` or so, fairly easy, you'd just add some code here gated on stage 2 I guess (or whatever logic you go with): https://github.com/rust-lang/rust/blob/2c462a2f776b899d46743b1b44eda976e846e61d/src/bootstrap/builder.rs#L772

File: src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs
Comment (anp) [reply src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs]: Cool. It occurs to me from a recent conversation at work that a blanket track_caller in traits would (I think?) allow error traits to automatically record callchain information with `?`. Exciting if true.

**Comment by anp** [comment]
Picking this back up, what do @eddyb and @Centril think about pivoting this PR to implement the attribute-on-trait-prototypes behavior and leaving the specific std annotations for a separate PR, kind of a mix of the suggestions in the subthreads on warn v error and perf runs.

**Comment by Centril** [comment]
I think it makes sense to provide the infrastructure first before applying it to specific impls and whatnot.

File: src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs
Comment (anp) [reply src/test/ui/rfc-2091-track-caller/warn-fn-prototype.rs]: https://github.com/rust-lang/reference/pull/742/files#r396035394 for reference docs, changing `warn-fn-prototype.rs` -> `error-extern-fn.rs`.

**Comment by Centril** [comment]
Reassigning to technical reviewer as lang team has given its blessing. r? @eddyb

**Comment by anp** [comment]
I split the perf-sensitive change out in https://github.com/rust-lang/rust/pull/70234.

Added tests here for the use of the attribute on trait method prototypes taking effect on impls (as discussed in review thread), now need to make it work.

File: src/librustc_typeck/collect.rs
Comment (eddyb) [inline src/librustc_typeck/collect.rs]: This shouldn't pass tidy.

Also, regarding the contents themselves, "implements a method" could be expanded to describe the fact that this is method in a trait impl.

File: src/librustc_typeck/collect.rs
Comment (eddyb) [inline src/librustc_typeck/collect.rs]: `impl_item.kind == ty::AssocKind::Method && ...`

File: src/librustc/ty/mod.rs
Comment (eddyb) [inline src/librustc/ty/mod.rs]: Now that this is fixed, maybe this special case can be removed: https://github.com/rust-lang/rust/blob/d1e81ef234ff5c2e0e3a69cb4e8e5f5b0fe1fd83/src/librustc/ty/mod.rs#L3057-L3063

File: src/librustc_passes/check_attr.rs
Comment (eddyb) [inline src/librustc_passes/check_attr.rs]: Is this language used elsewhere? I would've expected "on foreign functions".

File: src/librustc_typeck/collect.rs
Comment (eddyb) [inline src/librustc_typeck/collect.rs]: `container` -> `impl_def_id`

File: src/librustc_typeck/collect.rs
Comment (eddyb) [inline src/librustc_typeck/collect.rs]: Shouldn't need `.copied()`, AFAICT the element type is a `&'tcx ty::AssocItem`.

Comment (anp) [reply src/librustc_typeck/collect.rs]: it's my fault bc i `.copied()` from the example query in the specialization graph you showed me!

File: src/librustc_typeck/collect.rs
Comment (eddyb) [inline src/librustc_typeck/collect.rs]: `impl_item` -> `trait_item`

File: src/librustc_typeck/collect.rs
Comment (eddyb) [inline src/librustc_typeck/collect.rs]: `trait_method_id` -> `trait_item`

File: src/librustc_typeck/collect.rs
Comment (eddyb) [inline src/librustc_typeck/collect.rs]: `assoc` -> `impl_item` (unless you want to use some suffix other than `_item`)

File: src/librustc_typeck/collect.rs
Comment (eddyb) [inline src/librustc_typeck/collect.rs]: `trait_id` -> `trait_def_id`

File: src/librustc_typeck/collect.rs
Comment (eddyb) [inline src/librustc_typeck/collect.rs]: Order is `use_ident`, `def_ident`, `def_parent`.
So `impl_item.ident`, `trait_item.ident`, `trait_def_id`.

File: src/librustc_typeck/collect.rs
Comment (eddyb) [inline src/librustc_typeck/collect.rs]: I wonder if we should iterate over specialization ancestors, so that each of them can add `#[track_caller]`, not just the `trait` and the `fn` with a body.

File: src/test/ui/rfc-2091-track-caller/tracked-trait-impls.rs
Comment (eddyb) [inline src/test/ui/rfc-2091-track-caller/tracked-trait-impls.rs]: Can you add a second `impl` where you override this default method?

File: src/test/ui/rfc-2091-track-caller/tracked-trait-impls.rs
Comment (eddyb) [inline src/test/ui/rfc-2091-track-caller/tracked-trait-impls.rs]: This sort of thing could be a macro, but I don't care too much either way.

File: src/librustc_typeck/collect.rs
Comment (anp) [reply src/librustc_typeck/collect.rs]: That seems like the "most correct" option, unsure the implementation complexity is worth it. If you think it'll be easy can you give me some pointers? If not, I'll file a bug?

File: src/test/ui/rfc-2091-track-caller/tracked-trait-impls.rs
Comment (anp) [reply src/test/ui/rfc-2091-track-caller/tracked-trait-impls.rs]: There are two possible cases when overriding the default: with the attribute and without. Should both have the codegen flag?

Comment (anp) [reply src/test/ui/rfc-2091-track-caller/tracked-trait-impls.rs]: I'm going with "yes" since that's what happens by default with the codegen flag inheritance.

File: src/librustc_typeck/collect.rs
Comment (eddyb) [reply src/librustc_typeck/collect.rs]: Better to file an issue. Specialization is unstable anyway.

**Review by eddyb — APPROVED** [review]

File: src/librustc_typeck/collect.rs
Comment (anp) [reply src/librustc_typeck/collect.rs]: https://github.com/rust-lang/rust/issues/70293

**Pull request merged**
